File: clients/python/baml_lib/_impl/deserializer/primitive_deserializer.py
import typing

from .base_deserializer import (
    BaseDeserializer,
    CheckLutFn,
    RawWrapper,
    Result,
    Diagnostics,
)

T = typing.TypeVar("T", str, int, float, bool)


@typing.final
class PrimitiveDeserializer(BaseDeserializer[T]):
    __as_type: typing.Callable[[RawWrapper], typing.Optional[T]]
    __error_message: str

    def __init__(
        self,
        as_type: typing.Callable[[RawWrapper], typing.Optional[T]],
        error_message: str,
        rank: int,
    ):
        self.__as_type = as_type
        self.__error_message = error_message
        super().__init__(rank=rank)

    def coerce(
        self,
        raw: RawWrapper,
        diagnostics: Diagnostics,
        from_lut: CheckLutFn[T],
    ) -> Result[T]:
        parsed = self.__as_type(raw)
        if parsed is None:
            diagnostics.push_unknown_error(self.__error_message)
            return Result.failed()
        return Result.from_value(parsed)


@typing.final
class NoneDeserializer(BaseDeserializer[None]):
    def __init__(self) -> None:
        super().__init__(rank=0)

    def coerce(
        self,
        raw: RawWrapper,
        diagnostics: Diagnostics,
        from_lut: CheckLutFn[None],
    ) -> Result[None]:
        # We don't actually care about the value of raw, since we're just
        # returning None.
        return Result.from_value(None)

File: clients/ts/src/deserializer/primitive_deserializer.ts
import { BaseDeserializer, CheckLutFn, Result } from './base_deserializer';
import { Diagnostics } from './diagnostics';
import { RawWrapper } from './raw_wrapper/raw_wrapper';

type AsTypeFunction<T> = (raw: RawWrapper) => T | undefined;

class PrimitiveDeserializer<T extends string | number | boolean> extends BaseDeserializer<T> {
    private asType: AsTypeFunction<T>;
    private errorMessage: string;

    constructor(asType: AsTypeFunction<T>, errorMessage: string, rank: number) {
        super(rank);
        this.asType = asType;
        this.errorMessage = errorMessage;
    }

    coerce(raw: RawWrapper, diagnostics: Diagnostics, fromLut: CheckLutFn<T>): Result<T> {
        const parsed = this.asType(raw);
        if (parsed === undefined) {
            diagnostics.pushUnknownError(this.errorMessage);
            return Result.failed();
        }
        return Result.from_value(parsed);
    }
}

class NoneDeserializer extends BaseDeserializer<null> {
    constructor() {
        super(0);
    }

    coerce(raw: RawWrapper, diagnostics: Diagnostics, fromLut: CheckLutFn<null>): Result<null> {
        return Result.from_value(null);
    }
}

export { NoneDeserializer, PrimitiveDeserializer };